Implement adaption prompt from Llama-Adapter paper

yeoedward opened the pull request:
Issue #235 
Paper: https://arxiv.org/abs/2303.16199
Official Implementation: https://github.com/ZrrSkywalker/LLaMA-Adapter

This implementation is quite specific to Llama although in principle the same technique can be used for other transformer models. It is implemented by replacing some of the `LlamaAttention` modules with `AdaptedAttention` modules that wrap the original ones, but bypass the original `forward()` so that we can inject trainable tokens and a gate. I can add a notebook example in a subsequent PR when I have the time.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/peft/pr_268). All of your documentation changes will be reflected on that endpoint.

yeoedward commented:
Fixed `black` formatting and merge conflicts.

younesbelkada reviewed (commented):
Thanks a mile for the PR! I think it's in a great shape! 
Can you run the styling checks ?
```bash
make style && make quality
```

yeoedward commented:
CI test failures were caused by Llama dependencies not being available in earlier versions of the `transformers` library.
To ensure backwards compatibility, I've done the following:
1. Added import guards for llama classes from `transformers`. This will allow using `peft` to be used with earlier versions of `transformers` without breaking.
2. Skip unit tests that require those classes if the classes are not found. CI tests should pass.

yeoedward commented:
Rebased on main and added support for multi-adapters

pacman100 reviewed (commented):
This is awesome ✨! Great work @yeoedward 🔥. Thank you for blazing fast PR, adding multi-adapter support and adding tests. 

Overall comments:
1. The only drawback of the current implementation is being tightly coupled to the Llama model making it less generic. 
2. When I was thinking of the implementation of this method, I imagined `target_modules` pointing to the attention module, something like `self_attn` instead of `layers` (which we can get as parent module via `_get_submodules` method.
3. To make the implementation generic, having `attention_target_modules` as another config which is a `List[nn.Module]`.  The `AdaptedAttention` module's new forward should call the normal Attention layer's forward without any changes (so no modified_forward). Then use the `attention_target_modules` to get the `query_states`, `adapter_k` and `adapter_v`. Here, if there is only single fused_qkv as in gpt-2/bloom, just use `outputs.split(3, dim=2)` and do the necessary assignments.

pacman100 commented:
>  Recompute query_states, which will require importing the Llama-specific apply_rotary_pos_emb() (less duplication of code but with the perf penalty of recomputation).

I am leaning towards this as this would be generic and recomputation penalty would be minimal.
There can be a model to postprocessing function mapping which could be used for things like `apply_rotary_pos_emb `. Refer this code block for one such example:

https://github.com/huggingface/peft/blob/main/src/peft/peft_model.py#L246-L249

yeoedward commented:
@pacman100 I've made the changes you suggested and rebased on main.

winglian commented:
@yeoedward have you been able to validate the implementation results against the paper's results? I gave your branch a quick whirl with the same hyper parameters as the paper, but it definitely doesn't seem to converge as quickly.

yeoedward commented:
@winglian I have validated the implementation on smaller problems that run on a CPU but haven't tried reproducing the paper's results at full problem size. When you say it doesn't converge as quickly, are you referring to number of epochs trained? Happy to look into it if you can provide a notebook / script, although I don't have access to 8 x A100s.

pacman100 commented on src/peft/tuners/adaption_prompt.py:
```diff
@@ -0,0 +1,339 @@
+
+        # (bsz, num_heads, adapter_len, head_dim)
+        adapter_k = (
+            self.model.k_proj(self.adaption_prompt)
```
```suggestion
            key
```

pacman100 commented on src/peft/tuners/adaption_prompt.py:
```diff
@@ -0,0 +1,339 @@
+        )
+        # (bsz, num_heads, adapter_len, head_dim)
+        adapter_v = (
+            self.model.v_proj(self.adaption_prompt)
```
```suggestion
            value
```

pacman100 commented on src/peft/tuners/adaption_prompt.py:
```diff
@@ -0,0 +1,339 @@
+        # (bsz, q_len, num_heads * head_dim)
+        adapter_output = torch.matmul(scores, adapter_v).transpose(1, 2).reshape(bsz, q_len, -1)
+        # (bsz, q_len, hidden_size)
+        adapter_output = self.model.o_proj(adapter_output)
```
```suggestion
        if o_proj_layer is not None:
            adapter_output = getattr(self.model, o_proj_layer)(adapter_output)
```

pacman100 commented on src/peft/tuners/adaption_prompt.py:
```diff
@@ -0,0 +1,339 @@
+
+
+# Contains the config that is specific to a transformers model type.
+ModelTypeConfig = namedtuple("ModelTypeConfig", ["compute_query_states", "target_modules"])
```
```suggestion
ModelTypeConfig = namedtuple("ModelTypeConfig", ["compute_query_states", "target_modules", "k_proj_layer", "v_proj_layer", "o_proj_layer"])
```

pacman100 commented on src/peft/tuners/adaption_prompt.py:
```diff
@@ -0,0 +1,339 @@
+ModelTypeConfig = namedtuple("ModelTypeConfig", ["compute_query_states", "target_modules"])
+# Mapping of transformers model types to their specific configuration.
+TRANSFORMERS_MODEL_CONFIG = {
+    "llama": ModelTypeConfig(compute_query_states=llama_compute_query_states, target_modules="self_attn"),
```
```suggestion
    "llama": ModelTypeConfig(compute_query_states=llama_compute_query_states, target_modules="self_attn", k_proj_layer="k_proj", v_proj_layer="v_proj", o_proj_layer="o_proj"),
```

pacman100 commented on src/peft/tuners/adaption_prompt.py:
```diff
@@ -0,0 +1,339 @@
+        Args:
+            kwargs: See the original LlamaAttention module.
+        """
+        if kwargs.get("use_cache", False):
```
Just passing remaining outputs as is should do work

yeoedward replied:
My original intention was to avoid introducing additional complexity, but I think you're right it's less work than I thought. I've added support for it along with a simple unit test.

pacman100 reviewed (commented):
Hello @yeoedward, thank you for iterating. i have left comments

yeoedward commented:
@pacman100 I've made the suggested changes and rebased on main.

pacman100 reviewed (approved):
Thank you @yeoedward  for iterating 🤗. This is awesome and you have done a great job writing extensive tests to cover many aspects of Adaption Prompt implementation. Thank you so much for putting in all the effort, LGTM! 🚀

pacman100 commented:
Could you please run `make style` and `make quality` to fix the quality issues?

yeoedward commented:
@pacman100 I couldn't reproduce the style error (which was in a file unrelated to this PR), but I rebased on `main` and fixed some conflicts, which perhaps was the reason for the style checker failing. I've re-run the unit tests and quality checks, so hopefully this is good to go. Thanks for reviewing!

pacman100 commented:
Thank you @yeoedward! 🤗

tathagata-raha commented:
It took me a bit of time to find that the paper has been implemented, given that it is present in the list of the Supported methods. However, I haven't seen any documentation in the Tuners page [here](https://huggingface.co/docs/peft/package_reference/tuners)
